Adds :has() to experimental feature list

dipikabh opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->
#### Summary
<!-- ✍️ In a sentence or two, describe your changes -->

- `:has()` is supported in FF103 behind the preference `layout.css.has-selector.enabled`. This PR adds it to the experimental feature list.
- This PR also updates the [`:has()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:has) page to remove some now irrelevant text from the 'Description' and comments out some portions to revisit in the next round of rewrite for this page.

#### Supporting details
<!-- 🔗 Link to supporting information, such as bug trackers, source control, release notes, and vendor docs. -->
Bug tracker: https://bugzilla.mozilla.org/show_bug.cgi?id=1771896
Doc issue tracker: https://github.com/mdn/content/issues/19572

dipikabh committed:
Adds :has() to experimental feature list

dipikabh committed:
Comments out text to revisit

dipikabh committed:
Minor update

bsmth commented on files/en-us/mozilla/firefox/experimental_features/index.md:
```diff
@@ -187,7 +187,7 @@ This feature renders control characters (Unicode category Cc) other than _tab_ (
 </table>
 
-### Property: initial-letter
+### initial-letter property
```
A few of the titles need capitalization after rephrasing

```suggestion
### Initial-letter property
```

bsmth commented on files/en-us/mozilla/firefox/experimental_features/index.md:
```diff
@@ -267,7 +267,7 @@ Support for using a single {{cssxref("number")}} as a {{cssxref("ratio")}} when
 </table>
 
-### Property: backdrop-filter
+### backdrop-filter property
```
Same as above

```suggestion
### Backdrop-filter property
```

bsmth commented:
I initially added a few suggestions to capitalize the section headings, but I realize now they start with literal property names and lowercase looks more correct. What do you think of something like this:

```markdown
### Add initial-letter property
<!-- or -->
### Support initial-letter property
```

bsmth commented on files/en-us/web/css/_colon_has/index.md:
````diff
@@ -14,10 +14,12 @@ browser-compat: css.selectors.has
 /* Selects any <a>, as long as it has an <img> element directly inside it  */
 let test = document.querySelector('a:has(> img)');
 ```
+-->
````
why is this commented out?

dipikabh replied:
Removed the text. See the next comment.

bsmth commented on files/en-us/web/css/_colon_has/index.md:
```diff
@@ -27,7 +29,8 @@ let test = document.querySelector('a:has(> img)');
 
-The `:has()` pseudo-class takes a relative selector list as an argument. In earlier revisions of the CSS Selectors Level 4 specification, `:has` had a limitation that it couldn't be used within stylesheets. Instead, it could only be used with functions like {{domxref("document.querySelector()")}}; this was due to performance concerns. _This limitation has been removed_ because no browser implemen ...
+The `:has()` pseudo-class takes a relative selector list as an argument.
+<!-- Instead, it could only be used with functions like document.querySelector(); this was due to performance concerns. This limitation has been removed because no browser implemented it that way. -->
```
I think the comment can go, can't it?

bsmth replied:
Also I think this is the only place in the docs where we have:

```
<forgiving-relative-selector-list>
```

as syntax, I think this should be described here. What does 'forgiving' mean in this context?

Maybe one or two more sentences can go in the description or in the opening para about what it does. [This post ](https://webkit.org/blog/13096/css-has-pseudo-class/)might be a good reference.

dipikabh replied:
- I commented out the two portions because I wanted to revisit them in the next round of rewrite for this page. I've removed the comments now because I can get them from history.
- I have opened a separate issue https://github.com/mdn/content/issues/19972 to work on improving the :has() page, including the syntax and examples. Thanks for the reference! (Just noticed that Josh has also added the same link for reference in the issue 🙂)

bsmth replied:
>I've removed the comments now because I can get them from history
yeah I guess it makes sense. I still see them though, maybe there's not pushed changes? 🤔 

>I have opened a separate issue for improvements
OK good, tnx 👍🏻

bsmth commented:
From CI ->

* Looks like `{{domxref("Document.autoplayPolicy", "autoplayPolicy")}}` is broken / doesn't exist? Not sure why.
* `/en-US/docs/Web/CSS/animation-composition` doesn't exist
*  is `/en-US/docs/Web/API/Media_Streams_API` -> `/en-US/docs/Web/API/Streams_API` ?

dipikabh commented:
> From CI -> **Experimental features doc:**
> 
>     * Looks like `{{domxref("Document.autoplayPolicy", "autoplayPolicy")}}` is broken / doesn't exist? Not sure why.
> 
>     * `/en-US/docs/Web/CSS/animation-composition` doesn't exist
> 
>     * is `/en-US/docs/Web/API/Media_Streams_API` -> `/en-US/docs/Web/API/Streams_API` ?
> 
> 
> **:has()**
> 
>     * `{{HTMLElement("h1")}}` should be `/en-US/docs/Web/HTML/Element/Heading_Elements`

Thanks for pointing these out!

- Looks like there is no page for "autoplayPolicy". I've removed the macro.
- The [PR](https://github.com/mdn/content/pull/19848) for the new `animation-composition` page is soon to be merged.
- Fixed the broken link `/en-US/docs/Web/API/Media_Streams_API`.
    Should be `en-US/docs/Web/API/Media_Capture_and_Streams_API`.
- In prod, the macro `{{HTMLElement("h1")}}` seems to be working fine. Can you confirm - https://developer.mozilla.org/en-US/docs/Web/CSS/:has#examples?

bsmth commented:
>     * In prod, the macro `{{HTMLElement("h1")}}` seems to be working fine. Can you confirm - https://developer.mozilla.org/en-US/docs/Web/CSS/:has#examples?

Works okay for me, but CI complains that it's a "flaw". I wonder if it's better to link to `/en-US/docs/Web/HTML/Element/Heading_Elements` directly?

dipikabh commented:
> I initially added a few suggestions to capitalize the section headings, but I realize now they start with literal property names and lowercase looks more correct. What do you think of something like this:
> 
> ```gfm
> ### Add initial-letter property
> <!-- or -->
> ### Support initial-letter property
> ```

I've gone back-and-forth on these titles every time I touch this file :) 
- These titles were earlier of the type "Property: initial-letter" along with some other titles in the same section as phrases. If we do change the titles to something like "Support for initial-letter property", there'll be a repetition of the phrase "Support for" across (some) titles, which might seem redundant. And I did not want the title to start with 'The', so no 'The initial-letter property'. Maybe we can go back to "Property: initial-letter" format and let others be phrases. What do you think?

- Each section like HTML, SVG, APIs is updated/maintained by a different person. So it is even more difficult to ensure consistency in titles across sections.

dipikabh committed:
Fixes review comments

bsmth reviewed (approved):
Looks good, thanks 👍🏻

bsmth commented:
> From CI -> 

Looks like Markdownlint also doesn't like multiple blank lines, there's one failing check

dipikabh committed:
Removes extra line

dipikabh commented:
> > ```
> > * In prod, the macro `{{HTMLElement("h1")}}` seems to be working fine. Can you confirm - https://developer.mozilla.org/en-US/docs/Web/CSS/:has#examples?
> > ```
> 
> Works okay for me, but CI complains that it's a "flaw". I wonder if it's better to link to `/en-US/docs/Web/HTML/Element/Heading_Elements` directly?

I think I'll leave it as is. There are three other HTMLElement macro calls in the same section to `<a>`, `<img>`, and `<p>` and no error is being thrown for them.

bsmth committed:
Merge branch 'main' into experimental-has-pseudo-class

bsmth committed:
Merge branch 'main' into experimental-has-pseudo-class
